moveit_ros/planning/rdf_loader/src/rdf_loader.cpp

```diff
@@ -115,13 +126,107 @@ rdf_loader::RDFLoader::RDFLoader(TiXmlDocument *urdf_doc, TiXmlDocument *srdf_do
     srdf_.reset(new srdf::Model());
     if (!srdf_->initXml(*urdf_, srdf_doc))
     {
-      ROS_ERROR("Unable to parse SRDF");
+      ROS_ERROR_NAMED("rdf_loader", "Unable to parse SRDF");
       srdf_.reset();
     }
   }
   else
   {
-    ROS_ERROR("Unable to parse URDF");
+    ROS_ERROR_NAMED("rdf_loader", "Unable to parse URDF");
     urdf_.reset();
   }
 }
+
+bool rdf_loader::RDFLoader::isXacroFile(const std::string& path)
+{
+  // TODO: implement case-insensitive search
+  return path.find(".xacro") != std::string::npos;
+}
+
+bool rdf_loader::RDFLoader::loadFileToString(std::string& buffer, const std::string& path)
+{
+  if (path.empty())
+  {
+    ROS_ERROR_NAMED("rdf_loader", "Path is empty");
+    return false;
+  }
+
+  std::ifstream stream(path.c_str());
+  if (!stream.good())
+  {
+    ROS_ERROR_NAMED("rdf_loader", "Unable to load path");
+    return false;
+  }
+
+  // Load the file to a string using an efficient memory allocation technique
+  stream.seekg(0, std::ios::end);
+  buffer.reserve(stream.tellg());
+  stream.seekg(0, std::ios::beg);
+  buffer.assign((std::istreambuf_iterator<char>(stream)), std::istreambuf_iterator<char>());
+  stream.close();
+
+  return true;
+}
+
+bool rdf_loader::RDFLoader::loadXacroFileToString(std::string& buffer, const std::string& path, const std::vector<std::string>& xacro_args)
+{
+  if (path.empty())
+    return false;
+
+  std::string cmd = "rosrun xacro xacro --inorder ";
+  for (std::vector<std::string>::const_iterator it = xacro_args.begin(); it != xacro_args.end(); ++it)
+    cmd += *it + " ";
+  cmd += path;
+
+  FILE* pipe = popen(cmd.c_str(), "r");
+  if (!pipe)
+  {
+    ROS_ERROR_NAMED("rdf_loader", "Unable to load path");
+    return false;
+  }
+
+  char pipe_buffer[128];
+  while (!feof(pipe))
+  {
+    if (fgets(pipe_buffer, 128, pipe) != NULL)
+      buffer += pipe_buffer;
+  }
+  pclose(pipe);
+
+  return true;
+}
+
+bool rdf_loader::RDFLoader::loadXmlFileToString(std::string& buffer, const std::string& path, const std::vector<std::string>& xacro_args)
+{
+  if (isXacroFile(path))
+  {
+    return loadXacroFileToString(buffer, path, xacro_args);
+  }
+
+  return loadFileToString(buffer, path);
+}
+
+bool rdf_loader::RDFLoader::loadPkgFileToString(std::string& buffer, const std::string& package_name,
+                                                const std::string& relative_path,
+                                                const std::vector<std::string>& xacro_args)
+{
+  std::string package_path = ros::package::getPath(package_name);
+
+  if (package_path.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("rdf_loader", "ROS was unable to find the package name '" << package_name << "'");
+    return false;
+  }
+
+  boost::filesystem::path path(package_path);
+  // Use boost to cross-platform combine paths
+  path = path / relative_path;
+
+  if (!boost::filesystem::exists(path))
```

Comment:
Now that it's checked in the sub-functions I don't see the necessity to check here too :)